### PR TITLE
[WIP] New model syntax

### PR DESCRIPTION
**Pull request by mweststrate**
See #282 

Todo:
* [x] implement
* [x] verify types
* [x] update base tests
* [x] add tests for volatile state / private funcs etc
* [x] restory async
* [x] add tests for nested async
* [x] make sure async is properly bound
* [x] restore snapshot processing
* [x] example updates
* [x] document fluent api
* [x] address async generator syntax with es5 target & typings (see #279)
* [x] implement fluent type api with`actions(self => actions).props(propDeclaration).name("some name")`. Each step should produce a new type (not modifying the current one)
* [x] express compose using fluent api (or drop it?)
* [x] support old syntax or provide a codemod
* [x] restore all skipped tests
* [x] document babel setup for generators / update bookshop example
* [x] doc updates
* [x] update async-action docs
* [x] tutorial updates
* [x] document code mod
* [ ] update codesandbox examples (ASAP after merge)
* [ ] update getting started tutorial and related sandboxes (ASAP after merge)
* [ ] update mobx-state-tree deps in example package jsons (ASAP after merge)

### CODE REVIEW

File: src/types/complex-types/object.ts [hunk 1]
```diff
@@ -98,7 +98,15 @@ export class ObjectType<S, T> extends ComplexType<S, T> implements IModelType<S,
             // todo assert plain object
             if (actions && isPlainObject(actions)) {
                 Object.keys(actions).forEach(name => {
-                    addHiddenFinalProp(self, name, createActionInvoker(self, name, actions[name]))
+                    const action = actions[name]
+                    if ((action as any).isAsyncMSTAction === true)
```
Comment (mattiamanzati) [inline src/types/complex-types/object.ts]: Is this only to avoid async actions emit also a normal action?

**Comment by mweststrate** [comment]
Yep.

Op do 10 aug. 2017 09:13 schreef Mattia Manzati <notifications@github.com>:

**Comment by Bnaya** [comment]
Shouldn't this PR be in the 1.0 milestone?

**Comment by mweststrate** [comment]
Yes. I think the related issue is

Op do 10 aug. 2017 16:32 schreef Bnaya Peretz <notifications@github.com>:

File: src/types/complex-types/object.ts (hunk 1 above)
Comment (mattiamanzati) [inline src/types/complex-types/object.ts]: Uhm, what about having a modifier like "silent(...action...)" instead of specifically tie it to async actions?

**Comment by Bnaya** [comment]
So add it there ;) its seems so empty 
https://github.com/mobxjs/mobx-state-tree/pulls?q=is%3Aopen+is%3Apr+milestone%3A%221.0+-+initial+release%22

**Comment by mattiamanzati** [comment]
I completely agree and fallen in love with these last commits <3
As said I'd only change the fact that we should emit yield events only if promises are wrapped in task(() => promise()) to ensure that no side effects are performed on the tree and only a value is returned to ensure correct time travel replay :)

File: src/core/async.ts [hunk 2]
```diff
@@ -48,42 +48,56 @@ export function asyncAction(asyncAction: any): any {
     return createAsyncActionInvoker(asyncAction.name, asyncAction)
```
Comment (mattiamanzati) [inline src/core/async.ts]: Uhm, can we add an optional first parameter for the name? Not sure if most generator polyfills will retain the function name.

File: docs/getting-started.md [hunk 3]
```diff
@@ -178,14 +181,18 @@ const User = types.model({
...
+    function addTodo(id, name) {
+        self.todos.set(id, Todo.create({ name }))
     }
+
+    return {addTodo}
 })
 ```
 [View sample in playground](https://goo.gl/Sj17Nq)
 
+Please notice the use of "self". Self is the object being constructed when an instance of your model is created. Thanks to the self object instance actions are "this-free", allowing you to be sure that they are correclty bound.
```
Comment (praxxis) [inline docs/getting-started.md]: `correctly`

**Pull request merged**